packages/skeleton/src/components/project.ts

```diff
@@ -0,0 +1,462 @@
+import fs from 'fs'
+import path from 'path'
+
+import chalk from 'chalk'
+
+import { getPaths } from '@redwoodjs/internal/dist/index'
+
+import { getRootPath } from '../lib/path'
+
+import { extractCells } from './cell'
+import type { RedwoodCell } from './cell'
+import { RedwoodError, RedwoodWarning } from './diagnostic'
+import { extractDirectives } from './directive'
+import type { RedwoodDirective } from './directive'
+import { extractFunctions } from './function'
+import type { RedwoodFunction } from './function'
+import { extractLayouts } from './layout'
+import type { RedwoodLayout } from './layout'
+import { extractPages } from './page'
+import type { RedwoodPage } from './page'
+import { extractRouters } from './router'
+import type { RedwoodRouter } from './router'
+import { extractSDLs } from './sdl/sdl'
+import type { RedwoodSDL } from './sdl/sdl'
+import { extractServices } from './service/service'
+import type { RedwoodService } from './service/service'
+import { extractSides } from './side'
+import type { RedwoodSide } from './side'
+import { RedwoodSkeleton } from './skeleton'
+import { extractTOMLs } from './toml'
+import type { RedwoodTOML } from './toml'
+
+/**
+ * Used to enumerate either JS or TS project types
+ */
+export enum RedwoodProjectType {
+  TYPESCRIPT = 'typescript',
+  JAVASCRIPT = 'javascript',
+}
+
+export class RedwoodProject extends RedwoodSkeleton {
+  warnings: RedwoodWarning[] = []
+  errors: RedwoodError[] = []
+
+  readonly type: RedwoodProjectType
+
+  private cells?: RedwoodCell[] | undefined
+  private directives?: RedwoodDirective[] | undefined
+  private functions?: RedwoodFunction[] | undefined
+  private layouts?: RedwoodLayout[] | undefined
+  private pages?: RedwoodPage[] | undefined
+  private routers?: RedwoodRouter[] | undefined
+  private sdls?: RedwoodSDL[] | undefined
+  private services?: RedwoodService[] | undefined
+  private sides?: RedwoodSide[] | undefined
+  private tomls?: RedwoodTOML[] | undefined
```

Comment:
It's probably worth using JS's private field syntax (`#cell`, and when accessing, `this.#cell`) unless you have a good reason otherwise since it isn't one to one with TS's `private` keyword: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

Also, the `?` and `undefined` union look redundant to me; I'd just go with `?`

```suggestion
  #cells?: RedwoodCell[]
  #directives?: RedwoodDirective[]
  #functions?: RedwoodFunction[]
  #layouts?: RedwoodLayout[]
  #pages?: RedwoodPage[]
  #routers?: RedwoodRouter[]
  #sdls?: RedwoodSDL[]
  #services?: RedwoodService[]
  #sides?: RedwoodSide[]
  #tomls?: RedwoodTOML[]
```

Reply:
I'm sorry I'm late to the party here, but why not use TS's class syntax when in TS land? 
`private` gets you: Nicer syntax for defining, nicer syntax for accessing, faster performance (meh, most likely negligible) 
`#` gets you: Real privacy. Can't do `project['cells']`
So, yeah, technically/theoretically `#` is better. But practically I prefer `private`